mithril-aggregator/src/services/prover.rs

```diff
@@ -114,38 +109,51 @@ impl ProverService for MithrilProverService {
-        let transactions_to_prove = self
-            .get_transactions_by_hashes_with_block_range(transaction_hashes.to_vec())
+        // 1 - Compute the set of block ranges with transactions to prove
+        let block_ranges_transactions = self.get_block_ranges(transaction_hashes).await?;
+        let block_range_transactions = self
+            .get_all_transactions_for_block_ranges(&block_ranges_transactions)
             .await?;
 
-        // 2 - Compute Transactions Merkle Tree
-        let transactions = self.transaction_retriever.get_up_to(up_to).await?;
-        let mk_map = self.compute_merkle_map_from_transactions(transactions)?;
-
-        // 3 - Compute proof for each transaction to prove
-        let mut transaction_hashes_certified = vec![];
-        for (_block_range, transaction) in transactions_to_prove {
-            let mk_tree_node_transaction_hash: MKTreeNode =
-                transaction.transaction_hash.to_owned().into();
-            if mk_map
-                .compute_proof(&[mk_tree_node_transaction_hash])
-                .is_ok()
-            {
-                transaction_hashes_certified.push(transaction.transaction_hash);
-            }
+        // 2 - Compute block ranges sub Merkle trees
+        let mut mk_trees = BTreeMap::new();
+        for (block_range, transactions) in block_range_transactions {
+            let mk_tree = MKTree::new(
+                &transactions
+                    .iter()
+                    .map(|t| t.transaction_hash.clone())
+                    .collect::<Vec<_>>(),
+            )?;
+            mk_trees.insert(block_range, mk_tree);
         }
 
-        if !transaction_hashes_certified.is_empty() {
-            let mk_leaves: Vec<MKTreeNode> = transaction_hashes_certified
-                .iter()
-                .map(|h| h.to_owned().into())
-                .collect();
-            let mk_proof = mk_map.compute_proof(&mk_leaves)?;
-            let transactions_set_proof_batch =
-                CardanoTransactionsSetProof::new(transaction_hashes_certified, mk_proof);
+        // 3 - Compute block range roots Merkle map
+        let mut mk_map = self
+            .block_range_root_retriever
+            .compute_merkle_map_from_block_range_roots(up_to.immutable_file_number)
+            .await?;
+
+        // 4 - Enrich the Merkle map with the block ranges Merkle trees
+        for (block_range, mk_tree) in mk_trees {
+            mk_map.insert(block_range, mk_tree.into())?;
+        }
 
-            Ok(vec![transactions_set_proof_batch])
+        // 5 - Compute the proof for all transactions
+        if let Ok(mk_proof) = mk_map.compute_proof(transaction_hashes) {
+            let transaction_hashes_certified = transaction_hashes
+                .iter()
+                .filter_map(|hash| {
+                    mk_proof
+                        .contains(&hash.to_owned().into())
+                        .ok()
+                        .map(|_| hash.clone())
```

Comment:
If I'm not wrong you don't need the `ok()` since filter map can filters out error results as well.

Reply:
It's the `map` that applies to an `Option` not an `Iterator` so `filter_map` does not apply.

Reply:
You're right, looking at this iterator again I think that you can use the following syntax instead (it's a little shorter, using filter_map isn't really needed here):
```rust
let transaction_hashes_certified: Vec<TransactionHash> = transaction_hashes
                .iter()
                .filter(|hash| mk_proof.contains(&hash.as_str().into()).is_ok())
                .cloned()
                .collect();```

Reply:
Done